ocaml/xapi/pool_periodic_update_sync.ml

```diff
@@ -0,0 +1,182 @@
+  | `weekly ->
+      let days =
+        let today_wday = Ptime.weekday ~tz_offset_s utc_now |> weekday_to_int in
+        match today_wday < day_configed_int with
+        | true ->
+            day_configed_int - today_wday
+        | false ->
+            day_configed_int + 7 - today_wday
+      in
+      let span = Ptime.Span.of_d_ps (days, 0L) |> Option.get in
+      Ptime.add_span utc_start_of_today span |> Option.get
+
+let next_scheduled_datetime ~delay ~utc_now ~tz_offset_s =
+  let span = Ptime.Span.of_float_s delay |> Option.get in
+  let target = Ptime.add_span utc_now span |> Option.get in
+  let (y, m, d), ((hh, mm, ss), _) = Ptime.to_date_time ~tz_offset_s target in
+  (y, m, d, hh, mm, ss)
+
+let print_next_schedule ~delay ~utc_now ~tz_offset_s =
+  debug "[PeriodicUpdateSync] delay for next update sync: %f seconds" delay ;
+  let y, m, d, hh, mm, ss =
+    next_scheduled_datetime ~delay ~utc_now ~tz_offset_s
+  in
+  debug
+    "[PeriodicUpdateSync] next update sync scheduled at pool time: %d-%d-%d, \
+     %d:%d:%d "
+    y m d hh mm ss
+
+let update_sync_delay_for_next_schedule_internal ~utc_now
+    ~utc_start_of_next_sched_day ~seconds_in_a_day =
+  let random_span = Ptime.Span.of_float_s seconds_in_a_day |> Option.get in
+  let utc_next_schedule =
+    Ptime.add_span utc_start_of_next_sched_day random_span |> Option.get
+  in
+  let calc_delay =
+    Ptime.diff utc_next_schedule utc_now |> Ptime.Span.to_float_s
+  in
+  Float.max calc_delay update_sync_minimum_interval
+
+let update_sync_delay_for_next_schedule ~__context =
+  let frequency =
+    Db.Pool.get_update_sync_frequency ~__context
+      ~self:(Helpers.get_pool ~__context)
+  in
+  let day_configed =
+    Db.Pool.get_update_sync_day ~__context ~self:(Helpers.get_pool ~__context)
+  in
+  debug
+    "[PeriodicUpdateSync] update_sync_delay_for_next_schedule, frequency=%s, \
+     day_configed=%Ld"
+    (frequency_to_str ~frequency)
+    day_configed ;
+  let day_configed_int = Int64.to_int day_configed in
+  let utc_now = Ptime_clock.now () in
+  let tz_offset_s = Ptime_clock.current_tz_offset_s () |> Option.get in
+  let seconds_in_a_day = seconds_random_within_a_day () in
+  let utc_start_of_next_sched_day =
+    utc_start_of_next_scheduled_day ~utc_now ~tz_offset_s ~frequency
+      ~day_configed_int
+  in
+  let delay =
+    update_sync_delay_for_next_schedule_internal ~utc_now
+      ~utc_start_of_next_sched_day ~seconds_in_a_day
+  in
+  print_next_schedule ~delay ~utc_now ~tz_offset_s ;
+  delay
+
+let rec update_sync () =
+  Server_helpers.exec_with_new_task "periodic_update_sync" (fun __context ->
+      Helpers.call_api_functions ~__context (fun rpc session_id ->
+          ( try
+              ignore
+                (Client.Pool.sync_updates ~rpc ~session_id
+                   ~self:(Helpers.get_pool ~__context)
+                   ~force:false ~token:"" ~token_id:""
+                )
+            with _ ->
```

Comment:
Could the info/message in the exception be packed into the alert? I think it would be helpful.

Reply:
Added an <exception></exception> section for it.